controllers/awscluster_controller_unit_test.go

```diff
@@ -0,0 +1,632 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/gomega"
+	"golang.org/x/net/context"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/mock_services"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/conditions"
+)
+
+func TestAWSClusterReconciler_Reconcile(t *testing.T) {
+	testCases := []struct {
+		name         string
+		awsCluster   *infrav1.AWSCluster
+		ownerCluster *clusterv1.Cluster
+		expectError  bool
+	}{
+		{
+			name:        "Should not reconcile if owner reference is not set",
+			awsCluster:  &infrav1.AWSCluster{ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("aws-test-%s", util.RandomString(5)), Namespace: "default"}},
```

Comment:
I think we follow that in E2E tests not for integration tests, but if you have something to support this case, happy to change the namespace for each test.

Reply:
since we are using testenv in this test, so we should create objects for different test in different namespace, so that they are isolated from each other.

Reply:
To simplify, I will try to create in different namespace.

Reply:
Done